Topic/Connection statistic measurements

forouher opened the pull request:
These patches add an enhancement to roscpp and rospy that monitors some metrics (period, traffic, message delay, dropped messages) over every ROS connection and periodically sends out aggregated statistics data over a common topic /statistics.

The rationale behind this is that it is currently hard to measure these things without actually subscribing to the topics. E.g., if you want to know at which frequency your Kinect camera publishes images, you have to check with "rostopic hz /foo". But that will create quite a noticable load on the system and in any case might show a misleading number.  These patches add a method to measure those things in a lightweight way for the whole system.

I've extended rqt_graph to annotate the edges of the ROS graph with the statistics measured:

![stats_viewer](https://cloud.githubusercontent.com/assets/676900/2579279/776bbb50-b99f-11e3-9eb8-a5451d611c08.png)

For backward compatibility these statistics are disabled by default. They can be enabled via "rosparam set enable_statistics true".

I believe this enhancement might of useful to the ROS community at large. If there are any questions, concerns, remarks, feel free to share them.

cheers
Dariush

forouher committed:
Add topic/connection statistics to roscpp

This patch adds a module to roscpp that monitors some metrics
over every ROS connection (period, dropped messages, traffic, delay)
and periodically sends out aggregated statistics data over a common
topic /statistics. It is designed to be lightweight so it can be used
in "production" environments.

Statistics monitoring is disabled by default. It has to be enabled
explicitly using "rosparam set enable_statistics true".

forouher committed:
Add topic/connection statistics to rospy

This patch adds a module to roscpp that monitors some metrics
over every ROS connection (period, dropped messages, traffic, delay)
and periodically sends out aggregated statistics data over a common
topic /statistics. It is designed to be lightweight so it can be used
in "production" environments.

Statistics monitoring is disabled by default. It has to be enabled
explicitly using "rosparam set enable_statistics true".

ros-pull-request-builder commented:
Can one of the admins verify this patch?

dirk-thomas commented on clients/rospy/src/rospy/impl/statistics.py:
```diff
@@ -0,0 +1,303 @@
+
+try:
+    from cStringIO import StringIO #Python 2.x
+    import thread as _thread # Python 2
```
What is these two imports needed for? They seem to be unused.

dirk-thomas commented on clients/rospy/src/rospy/impl/statistics.py:
```diff
@@ -0,0 +1,303 @@
+    from cStringIO import StringIO #Python 2.x
+    import thread as _thread # Python 2
+    python3 = 0
+    def isstring(s):
```
This function seem to be unused either.

dirk-thomas commented:
Please update the code to conform to PEP8 (e.g. spaces instead of tabs, spaces around operators, spaces after commas etc.).

It would be great if the parameters / constraints like (window size etc.) could also be configured using parameters.

forouher committed:
rospy: remove superfluous imports

forouher committed:
rospy/roscpp: refactor TopicStatistics message

use Duration where possible. switch from logging variance
to logging standard deviation (stddev), as this better fits
the Duration data type.

forouher committed:
roscpp: don't use tabs

forouher committed:
rospy: adhere to PEP 8 + cleanup of comments

don't use tabs

forouher committed:
rospy: make variables configurable through parameters

forouher committed:
rospy: bugfix for issue introduced during refactoring

dividing a rospy.Duration by a float doesn't work as expected.

forouher committed:
roscpp: make variables configurable through parameters

forouher committed:
rospy: fix a bug that makes window resizing behave badly

dirk-thomas commented on clients/rospy/src/rospy/impl/tcpros_base.py:
```diff
@@ -728,7 +731,7 @@ def receive_loop(self, msgs_callback):
                         msgs = self.receive_once()
                         if not self.done and not is_shutdown():
-                            msgs_callback(msgs)
+                            msgs_callback(msgs,self)
```
Please add a space after the comma (PEP8).

dirk-thomas commented:
Please also update the C++ code to comply with the code style: e.g. add spaces after commas as well as add spaces around operators. Also keep curly braces on separate lines. See http://wiki.ros.org/CppStyleGuide for details.

dirk-thomas commented on clients/roscpp/src/libros/statistics.cpp:
```diff
@@ -0,0 +1,208 @@
+        ros::NodeHandle n("~");
+        // creating the publisher in the constructor results in a deadlock. so do it here.
+        pub_ = n.advertise<rosgraph_msgs::TopicStatistics>("/statistics",1);
+      }
```
The indentation of this block seems to be wrong.

forouher committed:
rospy: PEP8

forouher committed:
roscpp: adhere to code conventions

dirk-thomas commented:
@ros-pull-request-builder retest this please

tfoote commented on clients/rospy/src/rospy/impl/statistics.py:
```diff
@@ -0,0 +1,255 @@
+        # Those subscribers won't have a header. But as these subscribers are rather rare
+        # ("rostopic hz" is the only one I know of), I'm gonna ignore them.
+        if msg._has_header:
+            self.delay_list_.append(arrival_time - msg.header.stamp)
```
This is not valid. Data can be from the past.

tfoote replied:
Playing back a bag file is a canonical use case which could give you several months of "delay".

This is more a measure of age of data, and the data should be timestamped as when it was observed. If there's multiple stages in a pipeline this number will increase at each stage.

dirk-thomas commented:
Closing in favor of #398.
